File: 322/go/322.go
func coinChange(coins []int, amount int) int {
    // check 
    if amount == 0 {
        return 0;
    }

    // make dp
    dp := make([]int, amount+1)
    for i := 1; i <= amount; i++{
        dp[i] = -1
    }

    // state transition equation
    for i := 1; i < amount+1; i++{
        for _, c := range coins{
            if i >= c {
                if dp[i-c] > -1 {
                    if dp[i] > -1{
                        dp[i] = min(dp[i], 1+dp[i-c])
                    }else{
                        dp[i] = 1+dp[i-c]
                    }
                }
                
            }
        }
    }

    return dp[amount]
}

func min(a int, b int) int{
    if a > b {
        return b
    }
    return a
}

File: 322/py/322.py

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # make dp, 0 <= amount <= 10^4
        maxN = amount + 1
        dp = [(amount+1)]*(amount+1)

        dp[0] = 0

        for i in range(1, amount+1):
            for c in coins:
                if i - c >= 0 :
                    dp[i] = min(dp[i], dp[i-c] + 1)

        if dp[amount] == maxN:
            return -1

        return dp[amount]